api/src/main/java/org/terrakube/api/plugin/state/RemoteTfeService.java

```diff
@@ -981,34 +976,20 @@ RunsDataList getRunsQueue(String organizationName) {
 
     RunsDataList getWorkspaceRuns(String workspaceId) {
         RunsDataList runsDataList = new RunsDataList();
-        runsDataList.setData(new ArrayList());
+        runsDataList.setData(new ArrayList<RunsModel>());
         Optional<Workspace> workspaceList = workspaceRepository.findById(UUID.fromString(workspaceId));
-        if (workspaceList.isPresent()) {
-            Workspace workspace = workspaceList.get();
-            Optional<List<Job>> jobList = jobRepository.findAllByWorkspaceAndStatusNotInOrderByIdAsc(
-                    workspace,
-                    Arrays.asList(
-                            JobStatus.failed,
-                            JobStatus.completed,
-                            JobStatus.rejected,
-                            JobStatus.cancelled,
-                            JobStatus.approved));
-
-            int runWorkspace = 0;
-            if (jobList.isPresent()) {
-                for (Job job : jobList.get()) {
-                    log.info("Run Workspace {} job {}", runWorkspace, job.getId());
-                    Optional<RunsData> runsData = Optional.ofNullable(getRun(job.getId(), null));
-                    if (runsData.isPresent()) {
-                        RunsModel runsModel = runsData.get().getData();
-                        runsModel.getAttributes().put("position-in-queue", runWorkspace);
-                        runsDataList.getData().add(runsModel);
-                        runWorkspace = runWorkspace + 1;
-                    }
-
-                }
-            }
+        if (!workspaceList.isPresent()) {
+            return runsDataList;
         }
+        Workspace workspace = workspaceList.get();
+        jobRepository.findAllByWorkspaceAndStatusInOrderByIdDesc(workspace, Arrays.asList(
+                JobStatus.pending,
+                JobStatus.running,
+                JobStatus.queue)).forEach(job -> {
```

Comment:
you also need to add waiting for aproval here too

Reply:
This worth a bit discussion. Technically yes the next run should queue if the previous run(s) are pending on approval for apply.

TFC works differently than Terrakube. TFC doesn't have the idea of workflow as Terrakube, so it only really has two types of operations: Plan and Apply, hence it's easy to get the operation type and deduce whether the current run should be queued upon the previous job based on the operation type, basically if previous job type is apply, then waiting to be queued, if previous job is plan, then let it go.

https://github.com/hashicorp/terraform/blob/0ae6bc34c40f5a96b36c43c18f098d30df40ff0c/internal/cloud/backend_common.go#L136

Terrakube uses workflows which could potentially leads to a large number of jobs left in "WaitingForApproval" state, especially when used with webhook previously that doesn't a dedicated workflow/template, meaning the workspace and webhook could share the same "Plan->Approval->Apply" template. Yes, this situation improves once the webhook has its own dedicated template, but how do we deal with historical jobs that are in "WaitingForApproval" state? If we add "WaitingForApproval" state to the above `getWorkspaceRuns` function, all new jobs will be sitting to be queued indefinitely so long as there is a "WaitingForApproval" job in the history, no matter old or new. 

As far as I can see, there seems two ways to deal with this:

1. Force the customer to cancel all previous jobs that's in "WaitingForApproval" state, not too bad actually especially with webhook now using dedicated template.
2. Write a Liquid script to update all previous job in a upgrade
3. Do something smart in code, potentially
     1. A quartz job to cancel all "WaitingForApproval" jobs that are old than say 24 hours, (you might have that already in code somewhere)
     2. A function to check and cancel all previous "WaitingForApproval" jobs so long as a new "Apply" task is scheduled. 
     
I feel that item 2 is good for short-term, and option i in item 3 could potentially be adopted for long-term use. What do you think? @alfespa17

Reply:
I had some logic to cancel all jobs that have been running for 6 hours in the code in this part of the code

https://github.com/AzBuilder/terrakube/blob/9b45945f3e9a2b509a91b440052831af330519ba/api/src/main/java/org/terrakube/api/plugin/scheduler/ScheduleJob.java#L74

That code is no longer working now because in the past a quartz job was executed every 1 minute to validate the job status and continue the execution (terrakube was really slow during execution becuase of that), but I changed that logic some months ago so a quartz job is created dynamically when the Job entity is updated using the Elide hooks here:

https://github.com/AzBuilder/terrakube/blob/9b45945f3e9a2b509a91b440052831af330519ba/api/src/main/java/org/terrakube/api/rs/hooks/job/JobManageHook.java#L37

This is the reason why jobs in "waitingForApproval" state are no longer cancel after 6 hours...... maybe we will have to create some separete quartz job that fix it

So in theory once we fix that validation for jobs shouldn't be running for more that 6 hours

Reply:
By the way calculating the capacity based on the executor replicas could partially work because there is an option to have ephemeral executors like explained here but it is a good starting point.

https://docs.terrakube.io/getting-started/deployment/ephemeral-agents

Reply:
I added the logic to cancel long running jobs in a separate PR [here](https://github.com/AzBuilder/terrakube/pull/1289/files)

Reply:
Cool, I've added "WaitingForApproval" to the status query as suggested.

> That could be an option but I am not sure how the terraform CLI will handle that because if I remember correctly the CLI is only waiting for the user to answer yes/no to run the apply and it is not actively checking the job status

No, it's not and the Job will be printed to "Waiting to be queued", there should be documentation/knowledge some where to let customer know that you need to have all "WaitingForApproval" cancelled/finished. 

> I added the logic to cancel long running jobs in a separate PR [here](https://github.com/AzBuilder/terrakube/pull/1289/files)

That's great, thanks. I feel running the scheduled task every 5 mins is too frequent for jobs that will only be updated every 6 hours, even running the schedule task every 1 or 2 hours is helpful. As you have mentioned, it could make Terrakube slow, actually my team has complained many times about this, wondering whether some profiling should be done.

Additionally, correct me if I am wrong, the jobs sent over executor/runner hold the access token which only is valid for 1 or 2 hours,  out of the four job status, "pending" and "queued" jobs won't run successfully anyway?

Reply:
You are correct the job queue logic can be improved, I will try to think how to improved it in the following days, I already did some progress in the latest release triggering the job directly once the job status is updated instead of waiting for 1 minute, maybe I can improve it some more for 2.23.0.

Regarding the schedule to cancel long running jobs maybe it is better to add some parameter so anyone can customize the value.

For your last question the access token could be valid for 1 or 2 hours but API sent all the information to the executor each time it executes a template "step" for example in the following 

```yaml
flow:
  - name: "Plan"
    type: "terraformPlan"
    step: 100
  - name: "Apply"
    type: "terraformApply"
    step: 200
```

Terrakube will send all the information to the executor so it can execute the "plan" and mark the "plan" as completed, later it will send all of the information for the "apply" and the executor will mark the "apply" as completed.

In a high level the job status will be moving like this:

```
Pending (plan) -> Queue (plan) -> Running (plan) -> Pending (apply) -> Queue (apply)  -> Running (apply)  -> Pending (other step if need it) -> Completed (No other steps to execute)
```